### PR TITLE
P2553 implementation

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 1]
```diff
@@ -57,6 +57,8 @@ struct layout_stride;
 struct layout_left {
   template <class Extents>
   class mapping {
+  public:
+    using size_type = typename Extents::size_type;
```
Comment (dalg24) [inline include/experimental/__p0009_bits/layout_left.hpp:62]: Plz move all the public type member definitions here (`layout_type`, `extents_type`)

File: benchmarks/copy/copy_layout_stride.cpp [hunk 2]
```diff
@@ -113,115 +115,41 @@ void BM_MDSpan_Copy_2D_stride(benchmark::State& state, MDSpan, LayoutMapping map
...
-BENCHMARK_CAPTURE(
-  BM_MDSpan_Copy_2D_stride, size_dyn_fixed_100d_100d,
-  stdex::mdspan<int, stdex::extents<dyn, dyn>, stdex::layout_stride<dyn, 1>>(),
-  stdex::layout_stride<dyn, 1>::template mapping<stdex::extents<dyn, dyn>>(
-    stdex::extents<dyn, dyn>{100, 100},
-    // layout right
-    std::array<size_t, 1>{100}
-  )
-);
-
```
Comment (dalg24) [inline benchmarks/copy/copy_layout_stride.cpp]: Why did you remove these?

File: compilation_tests/ctest_constructor_sfinae.cpp [hunk 3]
```diff
@@ -54,92 +54,92 @@ namespace stdex = std::experimental;
 
 MDSPAN_STATIC_TEST(
     std::is_constructible<
-        stdex::extents<2, stdex::dynamic_extent>,
+        stdex::extents<size_t,2, stdex::dynamic_extent>,
```
Comment (dalg24) [inline compilation_tests/ctest_constructor_sfinae.cpp:57]: don't we use clang-format?

File: compilation_tests/ctest_trivially_copyable.cpp [hunk 4]
```diff
@@ -52,17 +52,10 @@ namespace stdex = std::experimental;
 //==============================================================================
 // <editor-fold desc="helper utilities"> {{{1
 
-MDSPAN_STATIC_TEST(
-  !std::is_base_of<stdex::extents<1, 2, 3>, stdex::detail::__partially_static_sizes<1, 2, 3>>::value
-);
-
-MDSPAN_STATIC_TEST(
-  !std::is_base_of<stdex::detail::__partially_static_sizes<1, 2, 3>, stdex::extents<1, 2, 3>>::value
```
Comment (dalg24) [inline compilation_tests/ctest_trivially_copyable.cpp:60]: Did you mean to remove?

File: tests/test_mdspan_conversion.cpp [hunk 5]
```diff
@@ -51,9 +51,9 @@ namespace stdex = std::experimental;
 
 TEST(TestMdspanConversionConst, test_mdspan_conversion_const) {
   std::array<double, 6> a{};
-  stdex::mdspan<double, stdex::extents<2, 3>> s(a.data());
+  stdex::mdspan<double, stdex::extents<uint32_t, 2, 3>> s(a.data());
```
Comment (dalg24) [inline tests/test_mdspan_conversion.cpp:54]: Any particular reason here and below to chose something else than `size_t`?

**Review by dalg24 — COMMENTED** [review]
Do you plan on adding any new test?

File: tests/test_mdspan_conversion.cpp (hunk 5 above)
Comment (crtrott) [reply tests/test_mdspan_conversion.cpp:54]: To test that it actually works if it is not the default value.

File: include/experimental/__p0009_bits/extents.hpp [hunk 6]
```diff
@@ -97,18 +103,20 @@ struct __extents_tag { };
...
 #endif
 {
 public:
 
-  using size_type = size_t;
+  using rank_type = size_t;
+  using size_type = ThisSizeType;
 
-  using __storage_t = detail::__partially_static_sizes_tagged<detail::__extents_tag, Extents...>;
+// internal typedefs which for technical reasons are public
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp:118]: Good comment (helpful for readers)

File: include/experimental/__p0009_bits/extents.hpp [hunk 7]
```diff
@@ -183,22 +191,24 @@ class extents
 public:
 
   MDSPAN_INLINE_FUNCTION
-  static constexpr size_t rank() noexcept { return sizeof...(Extents); }
+  static constexpr rank_type rank() noexcept { return sizeof...(Extents); }
   MDSPAN_INLINE_FUNCTION
-  static constexpr size_t rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((int(Extents == dynamic_extent)), /* + ... + */ 0); }
+  static constexpr rank_type rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((int(Extents == dynamic_extent)), /* + ... + */ 0); }
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]: Should we consider casting to `rank_type` inside, instead of to `int`, so as to avoid a potential conversion-to-unsigned warning?
```suggestion
  static constexpr rank_type rank_dynamic() noexcept { return _MDSPAN_FOLD_PLUS_RIGHT((rank_type(Extents == dynamic_extent)), /* + ... + */ 0); }
```

File: include/experimental/__p0009_bits/extents.hpp [hunk 8]
```diff
@@ -183,22 +191,24 @@ class extents
...
 
   //--------------------------------------------------------------------------------
   // Constructors, Destructors, and Assignment
 
   MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr extents() noexcept = default;
+#if !(MDSPAN_HAS_CXX20) && false
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]: This means `#if false`; should we consider just deleting this code?

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 9]
```diff
@@ -68,20 +71,43 @@ struct layout_left {
...
                  __extents.template __extent<r>() + i;
     }
 
     template<class I>
-    constexpr size_t __compute_offset(
+    constexpr size_type __compute_offset(
       __rank_count<Extents::rank()-1,Extents::rank()>, const I& i) const {
       return i;
     }
+#else
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_left.hpp]: Does this second option just exist for benchmarking?

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 10]
```diff
@@ -68,20 +71,43 @@ struct layout_left {
...
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_left.hpp]:
```suggestion
    // converted to int. That means users would need to choose index types
```

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 11]
```diff
@@ -68,20 +71,43 @@ struct layout_left {
...
+    // which he knows won't overflow i.e. the index type needs to be able
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_left.hpp]:
```suggestion
    // which they know won't overflow; i.e., the index type needs to be able
```

File: include/experimental/__p0009_bits/layout_left.hpp [hunk 12]
```diff
@@ -200,8 +225,8 @@ struct layout_left {
 #endif
 
     // Not really public, but currently needed to implement fully constexpr useable submdspan:
-    template<size_t N, size_t ... E, size_t ... Idx>
-    constexpr size_type __get_stride(std::experimental::extents<E...>,integer_sequence<size_t, Idx...>) const {
+    template<size_t N, class SizeType, size_t ... E, size_t ... Idx>
+    constexpr size_type __get_stride(std::experimental::extents<SizeType, E...>,integer_sequence<size_t, Idx...>) const {
       return _MDSPAN_FOLD_TIMES_RIGHT((Idx<N? __extents.template __extent<Idx>():1),1);
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_left.hpp:207]: Is the type of `__extent<Idx>()` always `size_t`?

File: include/experimental/__p0009_bits/layout_right.hpp [hunk 13]
```diff
@@ -68,24 +71,49 @@ struct layout_right {
...
       __rank_count<0,Extents::rank()>, const I& i, Indices... idx) const {
-      return __compute_offset(static_cast<size_t>(i),__rank_count<1,Extents::rank()>(),idx...);
+      return __compute_offset(i,__rank_count<1,Extents::rank()>(),idx...);
     }
 
-    constexpr size_t __compute_offset(size_t offset, __rank_count<Extents::rank(), Extents::rank()>) const {
+#else
+    // This option will cast __extents::extent() to the current index type.
+    // That means if you index into the mapping with just int everything gets
+    // converted to int. That means the user would need to choose index types
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_right.hpp]:
```suggestion
    // converted to int. That means users would need to choose index types
```

File: include/experimental/__p0009_bits/layout_right.hpp [hunk 14]
```diff
@@ -68,24 +71,49 @@ struct layout_right {
...
+    // which he knows won't overflow i.e. the index type needs to be able
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/layout_right.hpp]:
```suggestion
    // which they know won't overflow; i.e., the index type needs to be able
```

File: include/experimental/__p0009_bits/maybe_static_value.hpp [hunk 15]
```diff
@@ -61,18 +61,18 @@ namespace experimental {
 namespace detail {
 
 // static case
-template <class _T, _T __v,
-          _T __is_dynamic_sentinal = dynamic_extent,
+template <class _dynamic_t, class static_t, _static_t __v,
+          _static_t __is_dynamic_sentinal = dynamic_extent,
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/maybe_static_value.hpp:65]: Just a spelling error:
```suggestion
          _static_t __is_dynamic_sentinel = dynamic_extent,
```

File: include/experimental/__p0009_bits/maybe_static_value.hpp [hunk 16]
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
 };
 
 // dynamic case
-template <class _T, _T __is_dynamic_sentinal, size_t __array_entry_index>
-struct __maybe_static_value<_T, __is_dynamic_sentinal, __is_dynamic_sentinal,
+template <class _dynamic_t, class _static_t, _static_t __is_dynamic_sentinal, size_t __array_entry_index>
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/maybe_static_value.hpp:105]: (spelling error)
```suggestion
template <class _dynamic_t, class _static_t, _static_t __is_dynamic_sentinel, size_t __array_entry_index>
```

File: include/experimental/__p0009_bits/maybe_static_value.hpp [hunk 17]
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
...
+struct __maybe_static_value<_dynamic_t, _static_t, __is_dynamic_sentinal, __is_dynamic_sentinal,
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/maybe_static_value.hpp:106]:
```suggestion
struct __maybe_static_value<_dynamic_t, _static_t, __is_dynamic_sentinel, __is_dynamic_sentinel,
```

File: include/experimental/__p0009_bits/maybe_static_value.hpp [hunk 18]
```diff
@@ -102,20 +102,20 @@ struct __maybe_static_value {
...
                             __array_entry_index>
 #if !defined(_MDSPAN_USE_ATTRIBUTE_NO_UNIQUE_ADDRESS)
     : __no_unique_address_emulation<_T>
 #endif
 {
-  static constexpr _T __static_value = __is_dynamic_sentinal;
+  static constexpr _static_t __static_value = __is_dynamic_sentinal;
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/maybe_static_value.hpp:112]:
```suggestion
  static constexpr _static_t __static_value = __is_dynamic_sentinel;
```

File: include/experimental/__p0009_bits/mdspan.hpp [hunk 19]
```diff
@@ -335,20 +336,20 @@ MDSPAN_TEMPLATE_REQUIRES(
...
-  -> mdspan<ElementType, ::std::experimental::dextents<N>>;
+  -> mdspan<ElementType, ::std::experimental::dextents<size_t, N>>;
 
 // This one is necessary because all the constructors take `pointer`s, not
 // `ElementType*`s, and `pointer` is taken from `accessor_type::pointer`, which
 // seems to throw off automatic deduction guides.
-template <class ElementType, size_t... ExtentsPack>
-mdspan(ElementType*, const extents<ExtentsPack...>&)
-  -> mdspan<ElementType, ::std::experimental::extents<ExtentsPack...>>;
+template <class ElementType, class SizeType, size_t... ExtentsPack>
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/mdspan.hpp:350]: This one isn't in the proposal; do we need to add it?

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp [hunk 20]
```diff
@@ -88,29 +88,29 @@ template <size_t _I, class _T> using __repeated_with_idxs = _T;
  * @tparam _ValsSeq
  * @tparam __sentinal
  */
-template <class _Tag, class _T, class _ValsSeq, _T __sentinal = dynamic_extent,
+template <class _Tag, class _T, class _static_t, class _ValsSeq, _static_t __sentinal = static_cast<_static_t>(dynamic_extent),
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp:91]:
```suggestion
template <class _Tag, class _T, class _static_t, class _ValsSeq, _static_t __sentinel = static_cast<_static_t>(dynamic_extent),
```

Comment (mhoemmen) [reply include/experimental/__p0009_bits/standard_layout_static_array.hpp:91]: Generally, "sentinal" -> sentinel

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp [hunk 21]
```diff
@@ -141,8 +141,8 @@ struct __standard_layout_psa<
   // it's definition.
```
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp]:
```suggestion
  // its definition.
```

Comment (mhoemmen) [reply include/experimental/__p0009_bits/standard_layout_static_array.hpp]: also, `__size_dynamic` is mispelled ("`__size_synamic`") in the line above.

**Review by mhoemmen — DISMISSED** [review]
I did my best to review, but the innards really just need tests -- it's hard to tell with my eyeballs and brain what looks amiss.

File: include/experimental/__p0009_bits/mdspan.hpp (hunk 19 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/mdspan.hpp:350]: no we need to remove it, since it should work on its own (i.e. all template arguments are deducable?)

File: include/experimental/__p0009_bits/layout_left.hpp (hunk 9 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_left.hpp]: yeah I should remove it

File: include/experimental/__p0009_bits/extents.hpp (hunk 8 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: jup

File: include/experimental/__p0009_bits/extents.hpp (hunk 7 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: yeah one problem is I haven't systematically introduced rank_type yet

File: compilation_tests/ctest_constructor_sfinae.cpp (hunk 3 above)
Comment (crtrott) [reply compilation_tests/ctest_constructor_sfinae.cpp:57]: not yet :-(

File: compilation_tests/ctest_trivially_copyable.cpp (hunk 4 above)
Comment (crtrott) [reply compilation_tests/ctest_trivially_copyable.cpp:60]: probably shouldn't. :-)

Comment (crtrott) [reply compilation_tests/ctest_trivially_copyable.cpp:60]: I put it back in

File: include/experimental/__p0009_bits/extents.hpp (hunk 7 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: done

File: include/experimental/__p0009_bits/extents.hpp (hunk 8 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: I remove the if for now, and then we remove the constructors later in a harmoinzation PR.

File: include/experimental/__p0009_bits/layout_left.hpp (hunk 1 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_left.hpp:62]: will do later

File: include/experimental/__p0009_bits/layout_left.hpp (hunk 12 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_left.hpp:207]: no returns `size_type`, but that should be ok or? I mean `__get_stride` also returns `size_type`?

File: include/experimental/__p0009_bits/layout_right.hpp (hunk 13 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_right.hpp]: removed this all, since it was discussion of alternative options, the first ne is what we must DO

File: include/experimental/__p0009_bits/maybe_static_value.hpp (hunk 15 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/maybe_static_value.hpp:65]: I will search for all of those later.

File: include/experimental/__p0009_bits/mdspan.hpp (hunk 19 above)
Comment (crtrott) [reply include/experimental/__p0009_bits/mdspan.hpp:350]: but will do that as part of harmonization PR>

**Review by mhoemmen — APPROVED** [review]
